src/librbd/cache/pwl/rwl/replica/TimerPing.cc

```diff
@@ -0,0 +1,183 @@
+  }
+  freed_caches.clear();
+
+  if (has_need_free_cache) {
+    cls::rbd::RwlCacheDaemonNeedFreeCaches caches;
+    r = rwlcache_get_needfree_caches(&_io_ctx, _rados.get_instance_id(), caches);
+    if (r < 0) {
+      return r;
+    }
+    need_free_caches.swap(caches.need_free_caches);
+  }
+  return 0;
+}
+
+int DaemonPing::get_cache_info_from_filename(fs::path file, struct RwlCacheInfo& info) {
+  info.cache_size = fs::file_size(file);
+  std::string filename = file.filename();
+
+  if (filename.compare(0, 7, "rbd-pwl") != 0) {
+    return -1;
+  }
+
+  // find pool_name
+  std::string::size_type found = filename.find('.');
+  if (found == std::string::npos) {
+    return -1;
+  }
+  info.pool_name = filename.substr(found + 1, filename.find('.', found + 1) - found - 1);
+
+  // find image_name
+  found = filename.find('.', found + 1);
+  if (found == std::string::npos) {
+    return -1;
+  }
+  info.image_name = filename.substr(found + 1, filename.find('.', found + 1) - found - 1);
+
+  // find id
+  found = filename.rfind('.');
+  if (found == std::string::npos) {
+    return -1;
+  }
+  info.cache_id = std::strtoul(filename.substr(found + 1).c_str(), nullptr, 10);
+  return 0;
+}
+
+void DaemonPing::update_cacheinfos() {
+  std::string path = _cct->_conf->rwl_replica_path;
+
+  std::unordered_map<epoch_t, RwlCacheInfo> old_infos;
+  old_infos.swap(infos);
+  for (auto& p : fs::directory_iterator(path)) {
+    RwlCacheInfo info;
+    if (fs::is_regular_file(p.status())
+        && !get_cache_info_from_filename(p.path(), info)) {
+      infos.emplace(info.cache_id, std::move(info));
+      old_infos.erase(info.cache_id);
+    }
+  }
+  for (auto& info : old_infos) {
+    freed_caches.insert(info.first);
+  }
+}
+
+DaemonPing::C_Ping::C_Ping(std::shared_ptr<DaemonPing> dp) : dp(dp) {}
+DaemonPing::C_Ping::~C_Ping() {}
+void DaemonPing::C_Ping::finish(int r) {
+  int ret = dp->single_ping();
+  if (ret < 0) {
+    if (dp->_reactor) {
+      ldout(dp->_cct, 1) << "Error: shutdown in DaemonPing" << dendl;
+      dp->_reactor->shutdown();
+    }
+    return;
+  }
+
+  dp->_ping_timer.add_event_after(100, new C_Ping(dp));
+
+  dp->free_caches();
+}
+
+PrimaryPing::PrimaryPing(CephContext *cct, librados::IoCtx &io_ctx, ReplicaClient* client)
+  : _cct(cct), _io_ctx(io_ctx), _mutex(ceph::make_mutex("primary ping")),
+  _ping_timer(cct, _mutex), _client(client)  {
+    _ping_timer.init();
+}
+PrimaryPing::~PrimaryPing() {
+  ldout(_cct, 20) << dendl;
+  std::lock_guard locker(_mutex);
+  _ping_timer.cancel_all_events();
+  _ping_timer.shutdown();
+}
+
+int PrimaryPing::timer_ping() {
+  std::lock_guard locker(_mutex);
+  _ping_timer.add_event_after(5, new C_Ping(this));
+  return 0;
+}
+PrimaryPing::C_Ping::C_Ping(PrimaryPing* ping) : pping(ping) {}
+PrimaryPing::C_Ping::~C_Ping() {}
+void PrimaryPing::C_Ping::finish(int r) {
+  bool ok = pping->_client->single_ping();
+  if (ok) {
+    pping->_ping_timer.add_event_after(20, new C_Ping(pping));
```

Comment:
it should use RBD_RWLCACHE_PRIMARY_PING_TIMEOUT.

Reply:
done